statistics/handle/handle.go

```diff
@@ -1420,97 +1396,28 @@ func (h *Handle) TableStatsFromStorage(tableInfo *model.TableInfo, physicalID in
+	if err != nil {
 		return nil, err
 	}
-	table.ModifyCount = rows[0].GetInt64(0)
-	table.Count = rows[0].GetInt64(1)
-
-	rows, _, err = reader.Read("select table_id, is_index, hist_id, distinct_count, version, null_count, tot_col_size, stats_ver, flag, correlation, last_analyze_pos from mysql.stats_histograms where table_id = %?", physicalID)
-	// Check deleted table.
-	if err != nil || len(rows) == 0 {
-		return nil, nil
+	if reader.IsHistory() {
+		return statsTbl, err
 	}
-	for _, row := range rows {
-		if row.GetInt64(1) > 0 {
-			err = h.indexStatsFromStorage(reader, row, table, tableInfo)
-		} else {
-			err = h.columnStatsFromStorage(reader, row, table, tableInfo, loadAll)
-		}
-		if err != nil {
-			return nil, err
+	for histID, idx := range statsTbl.Indices {
+		if statistics.IsAnalyzed(idx.Flag) {
+			h.mu.rateMap.clear(physicalID, histID, true)
 		}
 	}
-	return h.extendedStatsFromStorage(reader, table, physicalID, loadAll)
-}
-
-func (h *Handle) extendedStatsFromStorage(reader *statistics.StatsReader, table *statistics.Table, physicalID int64, loadAll bool) (*statistics.Table, error) {
-	failpoint.Inject("injectExtStatsLoadErr", func() {
-		failpoint.Return(nil, errors.New("gofail extendedStatsFromStorage error"))
-	})
-	lastVersion := uint64(0)
-	if table.ExtendedStats != nil && !loadAll {
-		lastVersion = table.ExtendedStats.LastUpdateVersion
-	} else {
-		table.ExtendedStats = statistics.NewExtendedStatsColl()
-	}
-	rows, _, err := reader.Read("select name, status, type, column_ids, stats, version from mysql.stats_extended where table_id = %? and status in (%?, %?, %?) and version > %?", physicalID, StatsStatusInited, StatsStatusAnalyzed, StatsStatusDeleted, lastVersion)
-	if err != nil || len(rows) == 0 {
-		return table, nil
-	}
-	for _, row := range rows {
-		lastVersion = mathutil.Max(lastVersion, row.GetUint64(5))
-		name := row.GetString(0)
-		status := uint8(row.GetInt64(1))
-		if status == StatsStatusDeleted || status == StatsStatusInited {
-			delete(table.ExtendedStats.Stats, name)
-		} else {
-			item := &statistics.ExtendedStatsItem{
-				Tp: uint8(row.GetInt64(2)),
-			}
-			colIDs := row.GetString(3)
-			err := json.Unmarshal([]byte(colIDs), &item.ColIDs)
-			if err != nil {
-				logutil.BgLogger().Error("[stats] decode column IDs failed", zap.String("column_ids", colIDs), zap.Error(err))
-				return nil, err
-			}
-			statsStr := row.GetString(4)
-			if item.Tp == ast.StatsTypeCardinality || item.Tp == ast.StatsTypeCorrelation {
-				if statsStr != "" {
-					item.ScalarVals, err = strconv.ParseFloat(statsStr, 64)
-					if err != nil {
-						logutil.BgLogger().Error("[stats] parse scalar stats failed", zap.String("stats", statsStr), zap.Error(err))
-						return nil, err
-					}
-				}
-			} else {
-				item.StringVals = statsStr
-			}
-			table.ExtendedStats.Stats[name] = item
+	for histID, col := range statsTbl.Columns {
+		if statistics.IsAnalyzed(col.Flag) {
+			h.mu.rateMap.clear(physicalID, histID, false)
```

Comment:
Why move this part to here. Not in the origin function `columnStatsFromStorage` and `indexStatsFromStorage`

Reply:
Because we need `h.mu.rateMap` for the part.